server/sonar-server/src/main/java/org/sonar/server/permission/PermissionFinder.java

```diff
@@ -19,141 +19,75 @@
-    this.resourceDao = dbClient.resourceDao();
     this.permissionDao = dbClient.permissionDao();
+    this.userDao = dbClient.userDao();
+    this.groupDao = dbClient.groupDao();
   }
 
-  public List<UserWithPermissionDto> findUsersWithPermission(DbSession dbSession, PermissionQuery query) {
-    Long componentId = componentId(query.component());
-    int limit = query.pageSize();
-    return permissionDao.selectUsers(dbSession, query, componentId, offset(query), limit);
-  }
+  public List<UserDto> findUsers(DbSession dbSession, PermissionQuery.Builder dbQuery) {
+    List<String> orderedLogins = permissionDao.selectLoginsByPermissionQuery(dbSession, dbQuery.build());
 
-  /**
-   * Paging for groups search is done in Java in order to correctly handle the 'Anyone' group
-   */
-  public List<GroupWithPermissionDto> findGroupsWithPermission(DbSession dbSession, PermissionQuery query) {
-    Long componentId = componentId(query.component());
-    return toGroupQueryResult(permissionDao.selectGroups(dbSession, query, componentId), query);
+    return Ordering.explicit(orderedLogins).onResultOf(UserDto::getLogin).immutableSortedCopy(userDao.selectByLogins(dbSession, orderedLogins));
   }
 
-  @Nullable
-  private Long componentId(@Nullable String componentKey) {
-    if (componentKey == null) {
-      return null;
-    } else {
-      ResourceDto resourceDto = resourceDao.selectResource(ResourceQuery.create().setKey(componentKey));
-      if (resourceDto == null) {
-        throw new NotFoundException(String.format("Project '%s' does not exist", componentKey));
-      }
-      return resourceDto.getId();
+  public List<UserPermissionDto> findUserPermissions(DbSession dbSession, PermissionQuery.Builder dbQuery, List<UserDto> users) {
+    if (users.isEmpty()) {
+      return emptyList();
     }
-  }
-
-  private static List<GroupWithPermissionDto> toGroupQueryResult(List<GroupWithPermissionDto> dtos, PermissionQuery query) {
-    addAnyoneGroup(dtos, query);
-    List<GroupWithPermissionDto> filteredDtos = filterMembership(dtos, query);
-
-    Paging paging = forPageIndex(query.pageIndex())
-      .withPageSize(query.pageSize())
-      .andTotal(filteredDtos.size());
 
-    return pagedGroups(filteredDtos, paging);
+    List<String> logins = users.stream().map(UserDto::getLogin).collect(Collectors.toList());
+    return permissionDao.selectUserPermissionsByQuery(dbSession, dbQuery
+      .setLogins(logins)
+      .withPermissionOnly()
+      .build());
   }
 
-  private static int offset(PermissionQuery query) {
-    int pageSize = query.pageSize();
-    int pageIndex = query.pageIndex();
-    return (pageIndex - 1) * pageSize;
-  }
-
-  private static List<GroupWithPermissionDto> filterMembership(List<GroupWithPermissionDto> dtos, PermissionQuery query) {
-    return newArrayList(Iterables.filter(dtos, new GroupWithPermissionMatchQuery(query)));
-  }
+  public List<GroupDto> findGroups(DbSession dbSession, PermissionQuery.Builder dbQuery) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)